HPCC-11822 Improve archived workunit paging in WsWorkunits

wangkx opened the pull request:
1. Call getSortedDirectoryIterator() to retrieve the data about
archived workunits; 2. For archived workunits, the beforeWU/
afterWU and/or StartDate/EndDate are used to skip the directories
which store WUs not in the date range. The beforeWU/afterWU is
also used to skip the WUs which are not in the time range. 3. For
'online' workunits, the beforeWU/afterWU and/or StartDate/EndDate
are used to filter out the WUs which are not in the time range. 4. 
Remove PageStartFrom from paging request; 5. Refacter the code
inside WUiterate().

Also fix a bug: a '-' is needed for Before/After in SashaCommand.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-11822 Improve archived workunit paging in WsWorkunits

1. Call getSortedDirectoryIterator() to retrieve the data about
archived workunits; 2. For archived workunits, the beforeWU/
afterWU and/or StartDate/EndDate are used to skip the directories
which store WUs not in the date range. The beforeWU/afterWU is
also used to skip the WUs which are not in the time range. 3. For
'online' workunits, the beforeWU/afterWU and/or StartDate/EndDate
are used to filter out the WUs which are not in the time range.
4. Remove PageStartFrom from paging request; 5. Refacter the code
inside WUiterate().

Also fixed 2 bugs: 1. unsorted pages are returned by the WUiterate;
2. a '-' is needed for Before/After in SashaCommand.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 2f0213f592ea5669b3eee0cde9443e7bf2b829c8
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |117 |117 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |729 |729 |0 |
|test (thor) |642 |642 |0 |
|test (roxie) |756 |756 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented:
@jakesmith https://github.com/hpcc-systems/HPCC-Platform/pull/10497 is replaced by this PR. Please review.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                throw MakeStringException(-1,"SCA_WORKUNIT_SERVICES_GET not implemented for online workunits!");
+
+            Owned<IRemoteConnection> conn;
+            setSDSConnection(conn, mask);
```
minor: clearer if read:
```
Owned<IRemoteConnection> conn = getSDSConnection(mask);
```

wangkx replied:
Will be changed to getSDSConnection().

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            const char *owner = cmd->queryOwner();
+            const char *cluster = cmd->queryCluster();
+            const char *jobName = cmd->queryJobName();
+            const char *state = cmd->queryState();
```
Perhaps these should be members as checkFilters called often and these are also needed/used in other places, e.g. getWUs()

wangkx replied:
Ok, I will add them as members .

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+        }
+        bool checkFilters(IPropertyTree *t, bool onlineWU)
+        {
+            //The following code is a clone of the pre-existing code.
```
Which code is this comment referring to?

wangkx replied:
That is referring to the saarch.cpp, line 179-215 and 310 - 326.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                return false;
+
+            if (onlineWU)
+                return true;
```
I'm a bit confused by this, the workunit search dialog blanks out the ECL and logicalfile fields, if "Archive Only" is ticked.
Implying that you can use these fields for online workunits, but not archived workunits.

However, checkFilters with onlineWU=true, is used getOnlineWUs() and onlineWU=false for the archived version, so it's applying these filters to archived workunits only.

Is the search dialog wrong?
And can you only search by these fields in archived workunits?

wangkx replied:
I am using the onlineWU flag to simulate the old code (saarch.cpp, line 179-215 and 310 - 326). This part of the code is not used by ECLwatch's online WU search.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            const char *outputFormat = cmd->queryOutputFormat();
+            outputFields.appendList(outputFormat, "|,");
+        }
+        void getWUs()
```
This is dealing with archived workunits I think, would be better named getArchivedWUs

wangkx replied:
It will be named to getArchivedWUs.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
-                                continue;
-                            if (hasjobname&&(!pt.getProp("@jobName",val.clear())||!WildMatch(val.str(),jobname,true)))
+            setSDSConnection(conn, mask);
+            return isWild ? conn->queryRoot()->hasProp(wuid) : conn != nullptr;
```
perhaps should guard against conn being null here.

wangkx replied:
I will check the conn before using it.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            {
+                IPropertyTree *pt = &iter->query();
+                wuid.set(pt->queryName());
+                if (wuid.isEmpty())
```
Should be impossible, so curious why this check is here?

wangkx replied:
This check will be removed.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                }
+
+                if (sortInc)
+                {//The output should contain WUs from new to old. So, we need to reverse the order before output.
```
I don't think that's true here, the order coming back from getElements is not sorted.

wangkx replied:
Will add a sort function.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                if (wuCount == numWUs)
+                    break;
+            }
+            if (sortInc)
```
see comment above, I don't think your elements are sorted, so reversing won't help. You will instead need to actually sort them.

wangkx replied:
Will add a sort function.

jakesmith commented:
@wangkx - back to you.
Several of my questions may related to previous functionality, e.g. overflow leading 0 results.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 0a3022afd8eba8d1823aa3d65e326e0ca9614c9c
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   41 sec (00:00:41) |  169 sec (00:02:49) |   59 sec (00:00:59) |    6 sec (00:00:06) |   53 sec (00:00:53) | 1092 sec (00:18:12) |   38 sec (00:00:38) | 1458 sec (00:24:18) |

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+        Owned<IPropertyTree> wuTree;
+    };
+
+    class CArchivedWUReader : public CSimpleInterface
```
Ok, I will remove ' : public CSimpleInterface'.

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+        {
+            if (hasWUSOutput)
+            { //cmd->getAction()==SCA_WORKUNIT_SERVICES_GET
+                if (!serializeWUSrow(*wuTree, WUSbuf, false))
```
As we discussed, I will pass the WORKUNIT_SERVICES_BUFFER_MAX  into the serializeWUSrow().

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+
+    public:
+
+        CArchivedWUReader(ISashaCommand *_cmd, const char *_mask)
```
I will pass the cmd as a pointer.

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            checkBeforeOrAfterWU = !beforeWU.isEmpty() || !afterWU.isEmpty();
+            sortInc = !beforeWU.isEmpty() && afterWU.isEmpty();
+
+            numWUs = cmd->getLimit();
```
It will be changed to maxNumWUs.

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            else
+                getLdsPath("Archive/WorkUnits", path);
+
+            MemoryBuffer WUSbuf; //Not sure what this is.
```
The comment will be removed.

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                        if (outputModifiedTime)
+                            fileIterator->getModifiedTime(dt);
+
+                        if (!addOutput(outputFields, outputModifiedTime, hasWUSOutput, wuTree, dt, WUSbuf))
```
As we discussed, I will not change the existing logic here in this PR.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 6c49248b3767fe6116719506299cc633dd1e7738
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  166 sec (00:02:46) |   59 sec (00:00:59) |    7 sec (00:00:07) |   53 sec (00:00:53) | 1103 sec (00:18:23) |   40 sec (00:00:40) | 1452 sec (00:24:12) |

wangkx commented:
@jakesmith please review my changes.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 0868412af1e18452f61c37bbfe41c28d913d8beb
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  162 sec (00:02:42) |   60 sec (00:01:00) |    8 sec (00:00:08) |   55 sec (00:00:55) | 1110 sec (00:18:30) |   39 sec (00:00:39) | 1466 sec (00:24:26) |

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,439 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
-    if (haswusoutput)
-        cmd->setWUSresult(WUSbuf);
+        void getOnlineWUs()
+        { //The code here is refactered based on the pre-existing code.
```
trivial: don't think this comment will help clarify anyone revisiting the code, please remove.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,439 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            CIArrayOf<CWUData> wus;
+            unsigned wuCount = 0;
+            StringBuffer wuid;
+            Owned<IPropertyTreeIterator> iter = conn->queryRoot()->getElements(isWild ? "*" : NULL);
```
Not new, but shouldn't this getElements being passing the mask if isWild instead of "*"? 
i.e. get the property tree impl. to pre-filter - rather than get all and filter in parseAndCheckWUID with WildMatch

If so, could be a separate JIRA to address this.

wangkx replied:
We do not have to check the mask in both the getElements and the parseAndCheckWUID because the mask has been used in the getSDSConnection (line 469).

jakesmith commented:
@wangkx - a couple of comments, but could be merged.

Quite a big change, is there any kind of unit tests for this functionality?
Also it would be worth timing these requests on a Dali large'ish with a large # of workunits, to make sure it hasn't introduced a slowdown of some kind.

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 912d50b5a7f1e6d3fa656deda37a882a23a3da20
Number of Error(s): 10
 /usr/bin/install -c -m 644 include/antlr3.h include/antlr3baserecognizer.h include/antlr3basetree.h include/antlr3basetreeadaptor.h include/antlr3bitset.h include/antlr3collections.h include/antlr3commontoken.h include/antlr3commontree.h include/antlr3commontreeadaptor.h include/antlr3commontreenodestream.h include/antlr3convertutf.h include/antlr3cyclicdfa.h include/antlr3debugeventlistener.h include/antlr3defs.h include/antlr3encodings.h include/antlr3errors.h include/antlr3exception.h include/antlr3filestream.h include/antlr3input.h include/antlr3interfaces.h include/antlr3intstream.h include/antlr3lexer.h include/antlr3memory.h include/antlr3parser.h include/antlr3parsetree.h include/antlr3recognizersharedstate.h include/antlr3rewritestreams.h include/antlr3string.h include/antlr3tokenstream.h include/antlr3treeparser.h antlr3config.h /mnt/disk1/home/vamosax/smoketest/PR-10711/build/esp/services/ws_sql/antlr3c/include
collect2: error: ld returned 1 exit status
gmake[5]: *** [lib/libcouchbase.so.2.0.38] Error 1
gmake[4]: *** [CMakeFiles/couchbase.dir/all] Error 2
gmake[3]: *** [all] Error 2
make[2]: *** [plugins/couchbase/libcouchbase-build/lib/libcouchbase.so] Error 2
make[1]: *** [plugins/couchbase/CMakeFiles/libcouchbase-build.dir/all] Error 2
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result



[Build error:](https://gist.github.com/9b9104ce4514ffafee6bf92a6e3cc33f)
Build log [Build log](https://gist.github.com/HPCCSmoketest/9b9104ce4514ffafee6bf92a6e3cc33f#file-result-18-01-12-14-24-11-log)

AttilaVamos commented:
@wangkx It seems that was a network issue.

wangkx commented:
@jakesmith in https://github.com/hpcc-systems/HPCC-Platform/pull/10711/commits/d76822f9ed37dab3513f5a97c4ce239840eb28e1, I rebased my previous changes into one commit and rebased to the latest master.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 912d50b5a7f1e6d3fa656deda37a882a23a3da20
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   28 sec (00:00:28) |  175 sec (00:02:55) |   55 sec (00:00:55) |    6 sec (00:00:06) |   34 sec (00:00:34) | 1063 sec (00:17:43) |   32 sec (00:00:32) | 1393 sec (00:23:13) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 1699aabed304401c1a0032fc40401bdf871e2571
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  186 sec (00:03:06) |   55 sec (00:00:55) |    6 sec (00:00:06) |   35 sec (00:00:35) | 1099 sec (00:18:19) |   33 sec (00:00:33) | 1434 sec (00:23:54) |

wangkx commented:
@jakesmith the speed of retrieving online WUs is improved a lot by my last commit. Please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: e6e86e710d60a1543ff06876b11d96fcd8a28c47
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  181 sec (00:03:01) |   55 sec (00:00:55) |    5 sec (00:00:05) |   34 sec (00:00:34) | 1084 sec (00:18:04) |   33 sec (00:00:33) | 1410 sec (00:23:30) |

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,449 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+            IArrayOf<IPropertyTree> wus;
+            unsigned wuCount = 0;
+            StringBuffer wuid;
+            Owned<IPropertyTreeIterator> iter = conn->queryRoot()->getElements(isWild ? "*" : NULL);
```
```
jakesmith 5 days ago  Member
Not new, but shouldn't this getElements being passing the mask if isWild instead of "*"?
i.e. get the property tree impl. to pre-filter - rather than get all and filter in parseAndCheckWUID with WildMatch

If so, could be a separate JIRA to address this.
  @wangkx
wangkx 5 days ago  Member
We do not have to check the mask in both the getElements and the parseAndCheckWUID because the mask has been used in the getSDSConnection (line 469).
```

>because the mask has been used in the getSDSConnection (line 469).

There's something not right here..
It can't connect to the mask, e.g. connecting to /WorkUnits/W201701* will fail.

If a wild card is involved, it must connect to the root and filter.
And looking the current code (master) that's what it does. It only uses the mask if connect is it is *not* wild.

The same was supposed to be true here and it will fail with an getOnlineWUs call with a wild card as it is, afaics.
It should be : 
```
    Owned<IRemoteConnection> conn = getSDSConnection(isWild ? nullptr : mask.str());
```

Then, filtering _is_ needed .. and in the old code was done by WildMatch (line 306 in master), but could be done by getElements more directly.

wangkx replied:
Will fix.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,449 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                CriticalBlock block(wuDataSortSect);
+                wusRef = &wus;
+                wuDataSortInc = sortInc;
+                qsort(idx, nwus, sizeof(unsigned), compare);
```
globals and locks should be necessary.
You should be able to pass a sort compare function into the array::sort method directly.

wangkx replied:
Will use array::sort.

jakesmith commented:
@wangkx , assuming I'm right re. this issue (https://github.com/hpcc-systems/HPCC-Platform/pull/10711#discussion_r162131303), then it highlights the need for this is to be thoroughly tested before merging with some unit tests.

>I have no idea about the unit tests for sasha. Where is the existing unit testing code for sasha?

There aren't any at the moment, but I suspect the easiest way to construct a test for this and similar is via Esp soap requests and responses.
To test this change, e.g. to expose the above issue, need to create a few dummy workunits, then request some archived/online workunits with/without wildcards.

That won't be a trivial job.

For now, can you make sure that a variety of lookups to WUIterate work as before, e.g. with and without wildcards.

wangkx commented:
@jakesmith please review my changes.
I have two concerns for the unit tests of using ESP soap requests and responses. 1. how to create archived WUs through the existing ESP services? 2. the existing ESP services do not retrieve online WUs through sasha.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,446 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                }
+
+                pt->Link();
+                wus.append(*pt);
```
trivial: wua.append(*LINK(pt)) slightly tidier

wangkx replied:
Ok

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,446 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                CriticalBlock block(wuDataSortSect);
+                wusRef = &wus;
+                wuDataSortInc = sortInc;
+                idx.sort(compare);
```
As I say (here: https://github.com/hpcc-systems/HPCC-Platform/pull/10711#discussion_r162132190), globals and locks should not be necessary..

You don't need to build up a temporary array, set globals and use critical sections.
You can define a sort function and call wus.sort(suitableSortFunc) on it directly..

There are a few example that do same, in fact one in this file, see compareBranch, it's used by a array.bAdd() call instead of an array.sort() call, but the array and prototype for the CompareFunc are same for bAdd or sort.

If you look in jptree.cpp there's a IPropertyTree example, search fror comparePropTrees

jakesmith replied:
Are you sure the 'sortInc' ordering difference is correct here?
Surely you always want it in ascending order.. ?

The getArchivedWUs() branch uses sortInc to ensure that the array coming out of the reversed directory iterators is put back in ascending (old to new) order..
As you have it here currently, if sortInc is true, the sort function would sort it from new->old and return it like that..

wangkx replied:
> You can define a sort function and call wus.sort(suitableSortFunc) on it directly..
The wus is a tree array. The idx is an integer array. Will the wus.sort() be the same efficient as the idx.sort()?

jakesmith replied:
Yes, it will be more efficient (especially if concurrent request blocking on crit).
It's only sorting pointers, the current code will be performing lots of indirect calls to wus.item as well.

wangkx replied:
"Surely you always want it in ascending order.. ?"
The ideal order should be descending (from new to old). People may be more interested in newer WUs than older WUs. If so, for online WUs, I have to use the iptiter_sort first. Then, reverse and filter ... @jakesmith any suggestion?

jakesmith replied:
The line:
```
sortInc = !beforeWU.isEmpty() && afterWU.isEmpty();
```
.. could do with a comment explaining the purpose and reason "sortInc" is set in this case.

jakesmith replied:
>The ideal order should be descending (from new to old). People may be more interested in newer WUs than older WUs. If so, for online WUs, I have to use the iptiter_sort first. Then, reverse and filter ... @jakesmith any suggestion?

My main problem with the code as it is, is afaics it's inconsistent, it looks like:
```
Archived when sortInc=true, final order = old->new
Online when sortInc=true,  final order = old->new
Archived when sortInc=false, final order = old->new
Online when sortInc=false,  final order = *new->old*
```

i.e. it looks like the return order from archived is always old->new , is that not true?
And for online it looks like it is dependent on sortInc.

What order are they supposed to be returned in, in each case?
For the online case, if you want old->new order returned, using iptiter_sort would avoid the need for you to manually sort, if you need new->old order returned, then you'll have to sort yourself.
I don't think you need to change when the filtering is done.
@wangkx

wangkx replied:
The inconsistent sorting will be fixed. The sortInc may be renamed and will be commented.

wangkx replied:
I may add a client parameter (Descending) to control the order of returning WUs. The default order is new->old (Descending = true). The sortInc may be changed to countBackward. It is used to support paging. For example, a client retrieves archived WUs with the order new->old (Descending = true) and the client is on page 50. If a client wants to get the next page by specifying the afterWU, we may call the getSortedDirectoryIterator() with the order new->old, find the WU after the 'afterWU' and count a page of WUs. If a client wants to go back to the previous page by specifying the beforeWU, I may set countBackward = true (old->new), call the getSortedDirectoryIterator() with the order old->new, find the WU before the 'beforeWU', count a page of WUs, and reorder them to new->old. For onlineWUs, the logic may be similar but using iptiter_sort (for old->new) and manual sort (for new->old). It may take time to find the 'afterWU' or 'beforeWU' by iterating the sorted WUs. Similar logic will be implemented for Descending = false. @jakesmith any comment?

jakesmith replied:
@wangkx - I think that sounds fine.

jakesmith commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,446 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+
+static CriticalSection wuDataSortSect;
+static bool wuDataSortInc = true;
+static IArrayOf<IPropertyTree> *wusRef = nullptr;
```
These three global/statics should be removed.

wangkx replied:
ok

jakesmith commented:
@wangkx - see new comments.

Re. testing - proper unit tests for Esp/Sasha as I say are not trivial, so should open a new JIRA to come up with a approach to test this/other similar functionality systematically.

For now, need to ensure that it behaves as expected manually:
+ Test getArchivedWUs() and getOnlineWUs() functionality
+ make sure workunits that are archived, but restored (i.e. also online) do not get returned when archived only is set.
+ make sure order is correct, and remains so when paging backward and forward, and for getOnlineWUs
+ Make sure wildcard, single wuid and empy masks work as expected
+ Make sure the various filtering options work as expected/before.

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,446 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                cmd->addDT(modifiedTime);
+            return true;
+        }
+        bool addOutputFromWUTree(StringArray &outputFields, IPropertyTree *wuTree)
```
I should remove the outputFields from the input since it is a member variable.

wangkx commented on dali/sasha/saarch.cpp:
```diff
@@ -83,297 +83,446 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
-            conn.setown(querySDS().connect(xpath.str(), myProcessSession(), 0, 5*60*1000));
+            ForEachItemIn(i, outputFields)
+            {
+                const char* outputField = (const char*)outputFields.item(i);
```
ok

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 0750386e98be3d4aea8722969ee5646b01192921
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  181 sec (00:03:01) |   55 sec (00:00:55) |    6 sec (00:00:06) |   34 sec (00:00:34) | 1120 sec (00:18:40) |   33 sec (00:00:33) | 1459 sec (00:24:19) |

wangkx commented:
@jakesmith please review my last commit.

jakesmith commented:
@wangkx - looks good afaics.

Have you double checked / tested the above functionality list works as expected?
i.e.:
+ Test getArchivedWUs() and getOnlineWUs() functionality
+ make sure workunits that are archived, but restored (i.e. also online) do not get returned when archived only is set.
+ make sure order is correct, and remains so when paging backward and forward, and for getOnlineWUs
+ Make sure wildcard, single wuid and empy masks work as expected
+ Make sure the various filtering options work as expected/before.

Also, for some confidence that this has not degraded performance, can you get some timing stats of using this workunit paging before and after this PR in an environment with a decent amount of workunits?
I suspect it would be enough to run a trivial/very quick query a couple of thousand times to create that many workunits and then e.g. archive half of them via the sasha command line.

wangkx commented:
I did tested the functionality list. I will create a test environment with thousands of workunits...

wangkx commented:
@jakesmith I just emailed you a test stats.

jakesmith commented:
@wangkx - thanks replied to your email.
I would like to get to the bottom of the 4-7% speed decrease in fetching online workunits. If it is to do with the new sorting involved there, then is it necessary? 
The idea of the iptiter_sort was to *improve* efficiency, because it could avoid it iterating over all if a date range was supplied, sorting having ensured that the workunits were in date order.
But if you're page size is small and you're getting a random sample (unsorted) that fulfils the filter criteria.. then I can see how it could be quicker.
If the callers to WUiterate request online workunits didn't previously depend on ordering (?) and iptiter_sort is proved to be the cause of the 4%-7% slowdown - then there's no point in doing it that way.

jakesmith commented:
From an offline conversation, it looks like the online workunit retrieval is a bit slower because it now pages correctly, whereas before it was returning unsorted pages (sorted in esp) that were not sorted overall, essentially that was a bug.

So all looks good and optimization of paging archived workuntis (the main point of this PR) looks like a v. significant improvement.

jakesmith commented:
@wangkx - can you squash the commits and rebase and then ready to merge.

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 67ab08715a56f6144d04754dd5dcc04548a9b36f
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |82 |82 |0 |0 |0 |
|wutoolTest(Dali) |19 |17 |2 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |




Wutool error(s):
Dali:1) test: WuTool::testList (F) line: 1656 /mnt/disk1/home/vamosax/smoketest/PR-10505/HPCC-Platform/tools/wutool/wutool.cpp

Dali:equality assertion failed

Dali:2) test: WuTool::testSortByThorTime (F) line: 1930 /mnt/disk1/home/vamosax/smoketest/PR-10505/HPCC-Platform/tools/wutool/wutool.cpp

Dali:assertion failed


Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  178 sec (00:02:58) |   55 sec (00:00:55) |    0 sec (00:00:00) |   52 sec (00:00:52) | 1851 sec (00:30:51) |   62 sec (00:01:02) | 2207 sec (00:36:47) |

wangkx commented:
The same smoketest errors also exist in other PRs. @richardkchapman this PR is ready to be merged.

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: e77a60172ade657ca8f2c74847c53b0ce61248a5
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |82 |82 |0 |0 |0 |
|wutoolTest(Dali) |19 |17 |2 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |




Wutool error(s):
Dali:1) test: WuTool::testList (F) line: 1656 /mnt/disk1/home/vamosax/smoketest/PR-10505/HPCC-Platform/tools/wutool/wutool.cpp

Dali:equality assertion failed

Dali:2) test: WuTool::testSortByThorTime (F) line: 1930 /mnt/disk1/home/vamosax/smoketest/PR-10505/HPCC-Platform/tools/wutool/wutool.cpp

Dali:assertion failed


Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  184 sec (00:03:04) |   55 sec (00:00:55) |    0 sec (00:00:00) |   53 sec (00:00:53) | 1911 sec (00:31:51) |   62 sec (00:01:02) | 2270 sec (00:37:50) |
